_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+        |<p>Scan the QR Barcode attached and send us the token number.</p>
+        """.stripMargin
+
+    val file = Paths.get("barCode.png")
+    val content = new Content("text/html", message)
+
+    val mail = new Mail(from, subject, to, content)
+    val attachments = new Attachments()
+    attachments.setFilename(file.getFileName().toString())
+    attachments.setType("application/png")
+    attachments.setDisposition("attachment")
+    val attachmentContentBytes = Files.readAllBytes(file)
+    val attachmentContent = Base64.getEncoder().encodeToString(attachmentContentBytes)
+    attachments.setContent(attachmentContent)
+    mail.addAttachments(attachments)
+
+    val sg = new SendGrid(Properties.envOrElse("SENDGRID_API_KEY","undefined"))
+    val request = new Request()
+    val response: Try[Response] = Try {
+        request.setMethod(Method.POST)
+        request.setEndpoint("mail/send")
+        request.setBody(mail.build())
+        val res = sg.api(request)
+        res
+    }
+
+    val checkStatus = response match
+        case Success(res) =>
+            IO(println(res.getBody())) >>
+            IO(println(res.getStatusCode())) >>
+            IO(println(res.getHeaders())) >>
+            Ok("We sent you an email, please follow the steps to complete the signin process.")
+        case Failure(ex) =>
+            IO(println(ex.getMessage())) >>
+            Ok("Oops, something went wrong, please try again later.")
+end EmailService
+```
+
+### 3.1.4 Creating our FinalService
+
+Let's create a FinalService.scala file and add the following contents
+
+```scala
+package com.rockthejvm
+
+class FinalService(generator: String, id: Int):
+    lazy val otpService = new OtpService(generator, id)
+    lazy val emailService = new EmailService(otpService.getEmail(id))
+end FinalService
+
+```
+
+The `FinalService` class will be used to instantiate our `OtpService` and `EmailService` services with the appropriate constructor arguments. The `OtpService` takes its arguments from the `FinalService` constructor while the `EmailService` will use `optService` to acquire the email address of the user when provided with his or her `id`.
+
+### 3.1.5 Http4s Routes and Server
+
+In this section, we create the routes and server using http4s. Let's create a new scala file called OtpAuth.scala and add the following.
+
+```scala
+import cats.effect.{IOApp, IO, ExitCode}
+import com.xonal.FinalService
+
+object OtpAuth extends IOApp:
+    val tokenService = new FinalService("Hotp", 1)
+
+    override def run(args: List[String]): IO[ExitCode] =  ???
+```
+
+Here we create an instance of `FinalService` to which we supply a `generator` of `Hotp` and `id` of 1 as constructor arguments.
+
+```scala
+import org.http4s.*
+import org.http4s.dsl.io.*
+import scala.util.*
+    ...
+    val routes: HttpRoutes[IO] =
+        HttpRoutes.of[IO]{
+            case GET -> Root / "login" =>
+                tokenService.otpService.barCodeUrl match
```

Comment:
This is an opportunity to better integrate Cats Effect - you can make this barCodeUrl an IO[String], which you can then chain to the email sending, then the Ok you finally send to the caller

Reply:
Added IO to the services